fix: html entities decode error

xxkl1 opened the pull request:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] Prefix your PR title with `[feat]`, `[fix]`, `[chore]`, or `[docs]`.
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [x] Run the tests with `npm test` and lint the project with `npm run lint`

fixes : https://github.com/sveltejs/svelte/issues/8026

'&rect' be decoded to rectangle pattern, but '&rect' isn't html html escape entity.

```html
<!DOCTYPE html>
<html>

<head>
	<meta charset="utf-8" />
</head>

<body>
	<div>&amp</div>
	<div>&copy</div>
	<div>&rect</div>
	<hr>
	<div>&amp;</div>
	<div>&copy;</div>
	<div>&rect;</div>
	<hr>
	<div>&#169</div>
	<div>&#169;</div>
	<div>&#xA9</div>
	<div>&#xa9</div>
	<div>&#xA9;</div>
	<div>&#xa9;</div>
</body>

</html>
```
Firefox and Chrome render
<img width="94" alt="image" src="https://user-images.githubusercontent.com/84455605/203143703-b1f2a8e0-ab08-4de3-819f-19726f8bb265.png">

'&rect' is not be considered a html entity by browser.

The number html entity hasn't this problem looks.

This inspiration from [comment](https://github.com/sveltejs/svelte/issues/8026#issuecomment-1312682241) of @Prinzhorn 

Not only '&rect' has this problem, '&lstrok', '&lsquor', 'mho', etc such as. 

I have found a json file [entities.json](https://html.spec.whatwg.org/entities.json) from this [website](https://dev.w3.org/html5/html-author/charref) which in comments of [previous code](https://github.com/sveltejs/svelte/blob/master/src/compiler/parse/utils/entities.ts).

This json file not includes '&rect''&lstrok', '&lsquor', 'mho'.

I think it should be able to show that these strings are not html entities.

So these strings shouldn't be decoded theoretically.

The decode function of html entities should be fixed.

I have written a [little tool](https://stackblitz.com/edit/node-ph9jff?file=index.ts) to generate new entities.ts with message of semicolon by json file above.

And supplement unit tests for previous html decode function.

Ensure previous html decode funtion has unit tests fully.

And then add test for this issue, modify html decode funtion to past the test.

Now the html entities can be decode correctly and pasted all test successfully.

baseballyama commented on src/compiler/parse/utils/html.ts:
```diff
@@ -35,8 +35,18 @@ const windows_1252 = [
+	return String.raw`${entity_name}\b`;
+};
+
+const reg_exp_num = String.raw`#(?:x[a-fA-F\d]+|\d+)(?:;|\b)`;
```
THis is just a question.
Before the PR, it uses `\w`, but now `a-fA-F`.
Why do we need to handle only A to F?

xxkl1 replied:
I think we need use A to F and a to F. Because this regular expression only match hexadecimal String.Because of hexadecimal chars only A to F and a to f, It will occur match error if using \w.

eg, Using \w will match &#x9hello as hexadecimal html String and the  test unit 'hex number with other character 4'  will error.

baseballyama commented on test/utils/index.ts:
```diff
@@ -13,4 +15,309 @@ describe('utils', () => {
 		});
 	});
+
+	describe('decode_character_references', () => {
```
In terms of testing architecture, this file needs to move to `src/compiler/parse/utils/__test__.ts`.

xxkl1 replied:
Sorry, I have mistaken the positon of test unit.I will fix it.

baseballyama commented on test/utils/index.ts:
```diff
@@ -13,4 +15,309 @@ describe('utils', () => {
+				const value = decode_character_references(e);
+				const char = e.match(/\w+;?/)[0];
+				const codePoint = entities[char];
+				assert.equal(1, 1);
```
I think we don't need this.

xxkl1 replied:
Sorry, This is my carelessness.I will delele this in next commit.

baseballyama commented on test/utils/index.ts:
```diff
@@ -13,4 +15,309 @@ describe('utils', () => {
+			assert.equal(value, String.fromCodePoint(376));
+		});
+
+		it('validate code 3', () => {
+			// UTF-16 surrogate halves 55297 - 57343
+			const input = '&#57343;';
+			const value = decode_character_references(input);
+			assert.equal(value, String.fromCodePoint(0));
+		});
+
+		it('validate code 4', () => {
+			// out of max range
+			const input = '&#196608;';
+			const value = decode_character_references(input);
+			assert.equal(value, String.fromCodePoint(0));
+		});
```
If it renders with plain HTML, the result is different between HTML and Svelte.
Do we need to adjust them?

https://codepen.io/baseballyama/pen/bGKQwvz

xxkl1 replied:
I'm a bit unclear about the purpose of this processing.But it is old logic, i just fill test to it.I think it's best not to modify this logic.If there is a related issue by then, fix it.

xxkl1 commented on src/compiler/parse/utils/entities.ts:
```diff
@@ -1,2034 +1,2234 @@
-// https://dev.w3.org/html5/html-author/charref
+// https://html.spec.whatwg.org/entities.json from https://dev.w3.org/html5/html-author/charref
```
yes, you can search "This data is also available as a JSON file" in this page.

xxkl1 commented on src/compiler/parse/utils/entities.ts:
```diff
@@ -1,2034 +1,2234 @@
-	uwangle: 10663,
-	vzigzag: 10650,
-	zigrarr: 8669,
+	'AElig;': 198,
```
I think we should keep the order as previous PR. I will modify it in time.Thank for you suggestion.

xxkl1 commented:
Hello, I have fixed problem which discuss with you.

1.first commit:  modify [entities generate](https://stackblitz.com/edit/node-ph9jff) generate entities sort by key length and replace the `entities.ts`
2.second commit:  move the html decode test unit to `src/compiler/parse/utils/__test__.ts`
3.third commit: test following situation and keep it rendering same with plain html
```html
<div>&#x9hello</div>
<div>&#xf9hello</div>
<div>&#xF9hello</div>
<div>&#x9fhello</div>
<div>&#x9f;hello</div>
```

and then past all test and lint

can help review code and give some suggestion again for me if have time?
@baseballyama

Prinzhorn commented:
Maybe I've missed it, but could you add integration tests with the example from https://github.com/sveltejs/svelte/issues/8026? So that we make sure it actually fixes the issue with `&rect` inside an attribute?

1. `&rect` inside attribute stays literal `&rect`
2. `&rect;` inside attribute becomes &rect;
3. `&amp` (no semicolon) inside attribute becomes &amp;

baseballyama commented:
> Do you have any good suggestion on fixing this ci/test error？

Don't worry. CI is flaky.
I will make another PR for reducing the flakiness.

Prinzhorn commented:
> I have added test unit with example which you mentioned in `src/compiler/parse/utils/__test__.ts`

Thanks, I was mostly referring to testing attributes specifically. All tests are testing the entities inside HTML content, but not inside an attribute, where they behave different.

xxkl1 commented:
> > I have added test unit with example which you mentioned in `src/compiler/parse/utils/__test__.ts`
> 
> Thanks, I was mostly referring to testing attributes specifically. All tests are testing the entities inside HTML content, but not inside an attribute, where they behave different.

You are right.I'll add these tests in time.

Prinzhorn commented on src/compiler/parse/utils/__test__.ts:
```diff
@@ -298,6 +298,30 @@ describe('decode_character_references', () => {
+	it('escape character in attribute 3', () => {
+		// '&amp' decode to & which is html entity
+		const input = '<div style="background-image: url(https://url.com/path?lib=test&height=902&width=1500&amp=type1);"></div>';
+		const expect = '<div style="background-image: url(https://url.com/path?lib=test&height=902&width=1500&=type1);"></div>';
```
I don't think this is correct and the reason I asked for the tests. If you paste this into an HTML document the browser will make the following request (I recommend always using `example.com` as it's made for this exact purpose):

![image](https://user-images.githubusercontent.com/679144/205682106-7fb6348c-8d22-426a-b360-1f0ab047b130.png)

So inside attributes the semicolon appears to be mandatory. If you add the semicolon this happens:

![image](https://user-images.githubusercontent.com/679144/205682550-c22fc0bb-31c0-4a6f-9f98-71a721a47b09.png)

Prinzhorn replied:
I don't know enough about the Svelte internals to comment on that. But from what I understand the whole reason we need to do this is that Svelte uses `textContent` and [`setAttribute`](https://github.com/sveltejs/svelte/blob/cc52dc442be8cb3207005d13bf9b1d521b2b0afd/src/runtime/internal/dom.ts#L292-L295). So the HTML entities need to have been decoded already by the compiler. So if you don't decode the string for attributes it's still not correct. Because `setAttribute('data-test', '&amp;')` will result in the literal `&amp;` being the attribute. But `<div data-test="&amp;" />` is `&` in HTML. Parsing the serialized HTML is a different thing from working with the de-serialized DOM.

xxkl1 replied:
Yes, so it is no correct with no decode all attributes.

I think need to find out what needs to be decoded and what doesn't.

xxkl1 replied:
I don't know enough about the Svelte internals too, beacuse this is my first PR, haha.

I think you know enough better than me.👍

Prinzhorn replied:
Sounds good, let's wait for a maintainer to comment on everything

xxkl1 replied:
Ok,I going to start researching what needs to be decoded and what doesn't attribute.

And cases other than tags and content.

xxkl1 replied:
I'm going to ask baseballyama  to review tomorrow.

It's too late in Japan now, so I'm afraid to disturb him.

xxkl1 commented:
Hello, I have added test of style attributes and pass them.

new tests at `src/compiler/parse/utils/__test__.ts`:
`escape character in attribute 1`
`escape character in attribute 2`
`escape character in attribute 3`

Actually the best way is don't decode entities string which is attributes value.

By modify logic code `current_chunk.data = decode_character_references(current_chunk.raw)` at `src/compiler/parse/state/tag.ts`.

Because string of attribute value no decode in plain html.
https://codepen.io/xxkl1/pen/zYaywmE

However, there are possible risks in this way.

I think we can merge the current commit, becasue it can fixes the [issue](https://github.com/sveltejs/svelte/issues/8026) and keep low risk.

And then I make another PR to optimze the attributes value decode  and add the sufficient test for `src/compiler/parse/state/tag.ts` to ensure low risk.

@Prinzhorn

xxkl1 commented:
@baseballyama

xxkl1 commented:
@baseballyama
Hello, I have discussed [new problem](https://github.com/sveltejs/svelte/pull/8047#discussion_r1039780338) with @Prinzhorn yesterday.

This is our discussion summary
## The new problem:
https://codepen.io/xxkl1/pen/bGKOxdY
In this code pen, we can find html entity no need decode at html attribute style value.
Custom data attributes data-* value and html content need decode.
So test `escape character in attribute 3` in `src/compiler/parse/utils/__test__.ts` is not correct.

## How to fix this problem:
The result of my discussion with @Prinzhorn is that merge the current commit first, and then make new issue and PR to this new problem.

## Why need new issue and PR?
The reason is as follows
1.The problem is existing problems of previous commits.[reproduction](https://svelte.dev/repl/a06ecfc0c8424c4aa7ab868d0f63eac4?version=3.53.1)
2.The current PR can fixes [issue](https://github.com/sveltejs/svelte/issues/8026) which we need fix.
4.The [issue](https://github.com/sveltejs/svelte/issues/8026)'s title and reason is `Svelte compiler parses HTML entity without semicolon`, and the purpose of current PR is only fixing decode html entities without semicolon error too.
5.The purpose of fixing this new problem is fixing problem with what needs to be decoded and what doesn't decode in html.New purpose should be maked new issue to describe in detail and make new PR.
6.Solving this new problem requires more tests and modify more code which have possible risk.
7.It looks like most of the problems can be solved by merging the current commit with low risk.

## What to do next:
I have found out the entry of fixing this new problem is code `current_chunk.data = decode_character_references(current_chunk.raw)` at `src/compiler/parse/state/tag.ts`.

I going to add sufficient test for `src/compiler/parse/state/tag.ts` in new PR of new issue.
Find out what needs to be decoded and what doesn't decode in html with @Prinzhorn  in new issue discussion.
And then modify logic of code to fix this problem.

Can you help me review code and give me some suggestion for me if have time?
Thank you for always helping me and gave me a lot of good suggestion.

xxkl1 commented:
@baseballyama

xxkl1 commented:
Hello, I've optimized test of html entities decode.

Can you help me review code again for me if have time?

Thank you for always helping me and gave me a lot of good suggestion!
@baseballyama

Prinzhorn commented:
https://html.spec.whatwg.org/multipage/parsing.html#named-character-reference-state

> If the character reference was [consumed as part of an attribute](https://html.spec.whatwg.org/multipage/parsing.html#charref-in-attribute), and the last character matched is not a U+003B SEMICOLON character (;), and the [next input character](https://html.spec.whatwg.org/multipage/parsing.html#next-input-character) is either a U+003D EQUALS SIGN character (=) or an [ASCII alphanumeric](https://infra.spec.whatwg.org/#ascii-alphanumeric), then, for historical reasons, [flush code points consumed as a character reference](https://html.spec.whatwg.org/multipage/parsing.html#flush-code-points-consumed-as-a-character-reference) and switch to the [return state](https://html.spec.whatwg.org/multipage/parsing.html#return-state).

xxkl1 commented:
> https://html.spec.whatwg.org/multipage/parsing.html#named-character-reference-state
> 
> > If the character reference was [consumed as part of an attribute](https://html.spec.whatwg.org/multipage/parsing.html#charref-in-attribute), and the last character matched is not a U+003B SEMICOLON character (;), and the [next input character](https://html.spec.whatwg.org/multipage/parsing.html#next-input-character) is either a U+003D EQUALS SIGN character (=) or an [ASCII alphanumeric](https://infra.spec.whatwg.org/#ascii-alphanumeric), then, for historical reasons, [flush code points consumed as a character reference](https://html.spec.whatwg.org/multipage/parsing.html#flush-code-points-consumed-as-a-character-reference) and switch to the [return state](https://html.spec.whatwg.org/multipage/parsing.html#return-state).

As it says, I think about we shouldn't no encode  the html entity which not end withs ; and character is =, number or alphabet.

Prinzhorn commented:
> As it says, I think about we shouldn't no encode the html entity which not end withs ; and next character is =, number or alphabet.

_Maybe_ we should always require a semicolon? All these rules are for backwards compatibility and broken HTML. We're not trying to build a browser here, we could define that in a `*.svelte` file character references always need a semicolon. Svelte is not strictly HTML anyway, so we have this freedom. It would dramatically simplify this entire code. But that's not up for me to decide.

xxkl1 commented:
> > As it says, I think about we shouldn't no encode the html entity which not end withs ; and next character is =, number or alphabet.
> 
> _Maybe_ we should always require a semicolon? All these rules are for backwards compatibility and broken HTML. We're not trying to build a browser here, we could define that in a `*.svelte` file character references always need a semicolon. Svelte is not strictly HTML anyway, so we have this freedom. It would dramatically simplify this entire code. But that's not up for me to decide.

I think it's better to follow the browser's decode rules, otherwise there will be new issues of svelte.

xxkl1 commented:
Hello, friends, I have fixed all cases of html entities error which we have found, and optimzed test of html entities decode.

Cases we have found and plain html's render:
https://codepen.io/xxkl1/pen/vYrMGKg

Now the purpose of this PR is fixing all cases of html entities error!

Can help review my code and give some suggesstion for me if have time?
Hello, friends, I have fixed all cases of html entities error which we have found, and optimzed test of html entities decode.

Now the purpose of this PR is fixing all cases of html entities error!

Can help review my code and give some suggesstion for me if have time?
@baseballyama @Prinzhorn

tjkohli commented:
@xxkl1 @Prinzhorn and @baseballyama Congrats on 1.0! 🎉 

Thank you all for working on this issue.
Please let me know if I can help in any way. 🙏🏼

xxkl1 commented:
> 

I think you can review my commit😁

baseballyama commented:
Thank you for your update @xxkl1!
I will try to review the PR by next Friday.
I would appreciate it if @Prinzhorn could check it if you have time🙏

Prinzhorn commented:
I have one question regarding the general test setup, otherwise this lgtm. There are test files like `test/runtime/samples/html-entities/main.svelte` and the expected value in `_config.js`. I assume this is generated using the SSR code path and then compared with the serialized HTML? How is the client side code tested? Client side code will be something like

```js
span = element("span");
attr(span, "data-xxx", "▭=value");
attr(span, "style", "▭=value");
```

which will - at runtime - generated a DOM equivalent to the serialized SSR HTML.

So this is not directly related to this PR, I'm just wondering how the tests make sure that both code paths handle character references correctly? Or maybe this happens implicitly because the generated code is the same, it's just the implementation details of `element()` and `attr()` that are different in SSR?

baseballyama commented:
Each test runs 3 times. CSR / SSR / SSR and Hydration.
After rendering these, the test runner checks rendered HTML is matched with [html](https://github.com/sveltejs/svelte/pull/8047/files#diff-d791eb8a6dd21d88a52dbf76a1660293b6bfac40f60e550b1477d84ae708de3dR2).

If I don't answer to your question, please feel free to point it out.

Prinzhorn commented:
@baseballyama thanks, that's precisely what I asked

baseballyama reviewed (approved):
Thank you so much! LGTM!
This PR is a bit big, so I will wait for one more approval from another maintainer.

xxkl1 commented:
> Thank you so much! LGTM! This PR is a bit big, so I will wait for one more approval from another maintainer.

yes, I think so😂

dummdidumm reviewed (approved):
TLDR for others:
1. replace the big entities list with entities that have a `;` at the end where valid (there are some exceptions)
2. construct two regexes from that entities list: one for HTML where it's strictly matched, one for attributes where it tries to match browser behavior by also allowing `=` / digit / character afterwards
3. decode character references with one of these regexes depending on this is an attribute value or not

Code looks good to me
